cuML/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,93 @@
+# cuML developer guide
+This document summarizes rules and best practices for contributions to C++ component cuML of RAPIDS/cuml.
+
+# General
+
+# Thread safety
+
+# Coding style
+
+# Error handling
+All calls to CUDA APIs should be done via the provided helper macros `CUDA_CHECK`, `CUBLAS_CHECK` and `CUSOLVER_CHECK`. Those macros take care of checking the return values of the used API calls and generate an exception in case the command was no successful. In case an exception needs to be avoided, e.g. when implementing a destructor, `CUDA_CHECK_NO_THROW`, `CUBLAS_CHECK_NO_THROW ` and `CUSOLVER_CHECK_NO_THROW ` (currently not available, see https://github.com/rapidsai/cuml/issues/229) should be used. Those macros will only log the error but do not throw an exception.
+
+# Logging
+Add once https://github.com/rapidsai/cuml/issues/100 is addressed.
+
+# Device and Host memory allocations
+To enable `libcuml.so` users to control how memory for temporary data is allocated device memory should only be allocated via the allocator provided:
+```cpp
+template<typename T>
+void foo(ML::cumlHandle* handle, cudaStream_t stream, ... )
+{
+    T* temp_h = handle->getDeviceAllocator()->allocate(n*sizeof(T), stream);
+    ...
+    handle->getDeviceAllocator()->deallocate(temp_h, n*sizeof(T), stream);
+}
+```
+the same rule applies to larger amounts of host heap memory:
+```cpp
+template<typename T>
+void foo(ML::cumlHandle* handle, cudaStream_t stream, ... )
+{
+    T* temp_h = handle->getHostAllocator()->allocate(n*sizeof(T), stream);
+    ...
+    handle->getHostAllocator()->deallocate(temp_h, n*sizeof(T), stream);
+}
+```
+Small host memory heap allocations, e.g. as internally done by STL containers, are fine, e.g. an `std::vector` managing only a handful of integers.
```

Comment:
I have thought about that quite a bit. The advantage of passing in an allocator as a runtime argument over a template parameter are:
* Irrespective of the used allocator the containers remain type compatible.
* We can easily have multiple different allocator objects, e.g. one per device one per thread, .... 

With a template argument we would need to reference a global object. Does that make sense to you?

Reply:
All good points. 

I suppose since you are providing a host allocator and host_buffer, there's little reason for someone to use a `std::vector`. 

Another point in favor of "allocators as parameters" vs "template parameters" is that it allows stateful allocators, e.g., an allocator w/ a stream. This wouldn't be possible with an allocator as a template argument.